### PR TITLE
SILOptimizer: Allow inlining of transparent functions in `@backDeployed` thunks

### PR DESCRIPTION
**Pull request by tshortli**
In order for availability checks in iOS apps to be evaluated correctly when running on macOS, the application binary must call a copy of `_stdlib_isOSVersionAtLeast_AEIC()` that was emitted into the app, instead of calling the `_stdlib_isOSVersionAtLeast()` function provided by the standard library. This is because the call to the underlying compiler-rt function `__isPlatformVersionAtLeast()` must be given the correct platform identifier argument; if the call is not emitted into the client, then the macOS platform identifier is used and the iOS version number will be mistakenly interpreted as a macOS version number at runtime.

The `_stdlib_isOSVersionAtLeast()` function in the standard library is marked `@_transparent` on iOS so that its call to `_stdlib_isOSVersionAtLeast_AEIC()` is always inlined into the client. This works for the code generated by normal `if #available` checks, but for the `@backDeployed` function thunks, the calls to `_stdlib_isOSVersionAtLeast()` were not being inlined and that was causing calls to `@backDeployed` functions to crash in iOS apps running on macOS since their availability checks were being misevaluated.

The SIL optimizer has a heuristic which inhibits mandatory inlining in functions that are classified as thunks, in order to save code size. This heuristic needs to be relaxed in `@backDeployed` thunks, so that mandatory inlining of `_stdlib_isOSVersionAtLeast()` can behave as expected. The change should be safe since the only `@_transparent` function a `@backDeployed` thunk is ever expected to call is `_stdlib_isOSVersionAtLeast()`.
    
Resolves rdar://134793410.

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
@swift-ci please test Linux

**Comment by tshortli** [comment]
@swift-ci please test Windows

**Comment by tshortli** [comment]
@swift-ci please test

File: lib/SIL/IR/SILPrinter.cpp [hunk 1]
```diff
@@ -3387,6 +3387,7 @@ void SILFunction::print(SILPrintContext &PrintCtx) const {
 
   switch (isThunk()) {
   case IsNotThunk: break;
+  case IsBackDeployedThunk: // FIXME: Give this a distinct label
```
Comment (tshortli) [inline lib/SIL/IR/SILPrinter.cpp:3390]: I will do this in a subsequent PR. I didn't want the test churn because this change needs to be cherry-picked.

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
Run the benchmarks to be on the safe side (I don't expect any changes)

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
> Run the benchmarks to be on the safe side (I don't expect any changes)

We lost the link to these benchmark results since I had to rerun the tests after fixing some broken ones. Here are the original results: https://ci.swift.org/job/swift-PR-macos-perf/1001/

**Review by eeckstein — APPROVED** [review]
lgtm!

**Pull request merged**